platforms/core-configuration/kotlin-dsl-tooling-builders/src/main/kotlin/org/gradle/kotlin/dsl/tooling/builders/internal/IsolatedProjectsSafeKotlinDslScriptsModelBuilder.kt

```diff
@@ -0,0 +1,301 @@
+                additionalImports()
+            } ?: emptyList()
+
+        val classPath = (scriptClassPath + accessorsClassPath.bin)
+        val sourcePath = (gradleSource() + scriptSourcePath + accessorsClassPath.src)
+        return IntermediateScriptModel(scriptFile, classPath, sourcePath, additionalImports)
+    }
+
+    private
+    fun gradleSource() =
+        SourcePathProvider.sourcePathFor(
+            scriptClassPath,
+            scriptFile,
+            project.rootDir,
+            project.gradle.gradleHomeDir,
+            SourceDistributionResolver(project)
+        )
+}
+
+
+internal
+interface IsolatedScriptsModelParameter {
+    val parentClassPath: ClassPath
+    val parentSourcePath: ClassPath
+}
+
+
+internal
+data class DefaultIsolatedScriptModelParameter(
+    override val parentClassPath: ClassPath,
+    override val parentSourcePath: ClassPath
+) : IsolatedScriptsModelParameter
+
+
+internal
+data class IsolatedScriptsModel(
+    val models: List<IntermediateScriptModel>
+)
+
+
+internal
+class IsolatedScriptsModelBuilder(
+    private val intermediateModelProvider: IntermediateToolingModelProvider
+) : ParameterizedToolingModelBuilder<IsolatedScriptsModelParameter> {
+
+    override fun getParameterType() = IsolatedScriptsModelParameter::class.java
+
+    override fun canBuild(modelName: String): Boolean =
+        modelName == "org.gradle.kotlin.dsl.tooling.builders.internal.IsolatedScriptsModel"
+
+    override fun buildAll(modelName: String, project: Project): IsolatedScriptsModel {
+        error("Calling $this without parameter is not supported")
+    }
+
+    override fun buildAll(modelName: String, parameter: IsolatedScriptsModelParameter, project: Project): IsolatedScriptsModel {
+        val models = projectHierarchyScriptModels(project as ProjectInternal, parameter, intermediateModelProvider)
+        return IsolatedScriptsModel(models)
+    }
+}
+
+
+private
+fun projectHierarchyScriptModels(
+    project: ProjectInternal,
+    parameter: IsolatedScriptsModelParameter,
+    intermediateModelProvider: IntermediateToolingModelProvider
+): List<IntermediateScriptModel> {
+
+    val models = mutableListOf<IntermediateScriptModel>()
+
+    // TODO: compute own classpaths
+//    scriptClassPath = project.scriptCompilationClassPath,
+//    scriptSourcePath = sourcePathFor(sourceLookupScriptHandlersFor(project))
+    val ownClassPath = parameter.parentClassPath
+    val ownSourcePath = parameter.parentSourcePath
+
+    models += listOfNotNull(project.discoverBuildScript()?.let {
+        buildScriptModelBuilder(it, project, ownClassPath, ownSourcePath).buildModel()
+    })
+
+    models += project.discoverPrecompiledScriptPluginScripts().map {
+        // TODO:isolated support fully
+        IntermediateScriptModel(it, parameter.parentClassPath, parameter.parentSourcePath, emptyList())
+    }
+
+    val childrenParameter = DefaultIsolatedScriptModelParameter(ownClassPath, ownSourcePath)
```

Comment:
🤔 Passing the state down to the children has the downside that each cached child model will be storing the same information over and over (which will be compounded at each level of the project path tree). 

Instead I think the parent should prepend its own paths to each child. And this operation could be made very efficient by making the `IntermediateScriptModel`s carry `PersistentList<ClassPath>`s so the prepends at each level are `O(1)`.

Reply:
Even better would be to only aggregate at the top (at the root model builder level) and keep the cached intermediate models completely isolated. That should reduce IO considerably for project trees with height > 1.